src/hardware/mixer.cpp

```diff
@@ -406,160 +405,169 @@ void MixerChannel::AddSamples(uint16_t len, const Type *data)
-			pos++;
-#if MIXER_UPRAMP_STEPS > 0
-			if (last_samples_were_silence && pos == 1) {
-				offset[0] = next_sample[0] - prev_sample[0];
-				if (stereo) offset[1] = next_sample[1] - prev_sample[1];
-				//Don't bother with small steps.
-				if (offset[0] < (MIXER_UPRAMP_SAVE*4) && offset[0] > (-MIXER_UPRAMP_SAVE*4)) offset[0] = 0;
-				if (offset[1] < (MIXER_UPRAMP_SAVE*4) && offset[1] > (-MIXER_UPRAMP_SAVE*4)) offset[1] = 0;
-			}
-
-			if (offset[0] || offset[1]) {
-				next_sample[0] = next_sample[0] - (offset[0]*(MIXER_UPRAMP_STEPS*static_cast<Bits>(len)-static_cast<Bits>(pos))) /( MIXER_UPRAMP_STEPS*static_cast<Bits>(len) );
-				next_sample[1] = next_sample[1] - (offset[1]*(MIXER_UPRAMP_STEPS*static_cast<Bits>(len)-static_cast<Bits>(pos))) /( MIXER_UPRAMP_STEPS*static_cast<Bits>(len) );
-			}
-#endif
 		}
 
 		// Process initial samples through an expanding envelope to
 		// prevent severe clicks and pops. Becomes a no-op when done.
-		envelope.Process(stereo, interpolate, prev_sample, next_sample);
+		envelope.Process(stereo, prev_sample);
 
-		//Where to write
-		mixpos &= MIXER_BUFMASK;
-		if (!interpolate) {
+		const auto left = prev_sample[mapped_channel_left] * volmul[0];
+		const auto right = (stereo ? prev_sample[mapped_channel_right]
+		                           : prev_sample[mapped_channel_left]) *
+		                   volmul[1];
 
-			mixer.work[mixpos][mapped_output_left] +=
-			        prev_sample[mapped_channel_left] * volmul[0];
+		out_frame = {0.0f, 0.0f};
+		out_frame[mapped_output_left] += left;
+		out_frame[mapped_output_right] += right;
 
-			mixer.work[mixpos][mapped_output_right] +=
-			        (stereo ? prev_sample[mapped_channel_right]
-			                : prev_sample[mapped_channel_left]) *
-			        volmul[1];
+		out.push_back(out_frame[0]);
+		out.push_back(out_frame[1]);
 
-		} else {
-			const auto diff_mul = freq_counter & FREQ_MASK;
-			auto sample = prev_sample[mapped_channel_left] +
-			              (((next_sample[mapped_channel_left] - prev_sample[mapped_channel_left]) *
-			                diff_mul) >>
-			               FREQ_SHIFT);
-			mixer.work[mixpos][mapped_output_left] += sample * volmul[0];
-			if (stereo) {
-				sample = prev_sample[mapped_channel_right] +
-				         (((next_sample[mapped_channel_right] - prev_sample[mapped_channel_right]) *
-				           diff_mul) >>
-				          FREQ_SHIFT);
-			}
-			mixer.work[mixpos][mapped_output_right] += sample * volmul[1];
-		}
-		//Prepare for next sample
-		freq_counter += freq_add;
-		mixpos++;
-		done++;
+		++pos;
 	}
-	
+}
+
+spx_uint32_t estimate_max_out_frames(SpeexResamplerState *resampler,
+                                     const spx_uint32_t in_frames)
+{
+	assert(resampler);
+	assert(in_frames);
+
+	spx_uint32_t ratio_num = 0;
+	spx_uint32_t ratio_den = 0;
+	speex_resampler_get_ratio(resampler, &ratio_num, &ratio_den);
+	assert(ratio_num && ratio_den);
+
+	return ceil_udivide(in_frames * ratio_den, ratio_num);
```

Comment:
I original thought we have similar information:
  - `mixer.freq` should be `ratio_den`
  - the channel's `sample_rate` should be `ratio_num`

So thought perhaps we don't have to ask speex for stuff we already know.  But I understand Speex bakes out the common factors from the ratio, which can help avoid rolling over the 32-bit calculation (especially if the numbers got pretty large.. say 400,000 Hz in-rate * 16,000 samples).

(not asking for change.. just a thought!)

Reply:
Yeah, it's a bit safer, and I think I like it better that we ask the resampler to give us the numbers, so it's guaranteed we're using the correct numbers it's configured for.

Reply:
> it's guaranteed we're using the correct numbers it's configured for.

I like that reason; get details directly from the source.